spec/lang/representation.md

```diff
@@ -229,16 +229,54 @@ impl Type {
 
 ### Enums
 
-TODO: implement Enum decoding & encoding.
+Enum encoding and decoding.
+Note that the discriminant may not be written into bytes that contain encoded data.
+This is to ensure that pointers to the data always contain valid values.
 
 ```rust
+/// Computes the discriminant from the value's bytes by the decision tree in the `Discriminator` by
+/// recursing over the branches defined by `Discriminator::Unknown`.
+/// FIXME: we have multiple quite different fail sources, it would be nice to return more error information.
+fn compute_discriminant<M: Memory>(bytes: List<AbstractByte<M::Provenance>>, discriminator: Discriminator) -> Option<Int> {
+    match discriminator {
+        Discriminator::Known(val) => Some(val),
+        Discriminator::Invalid => None,
+        Discriminator::Unknown { offset, children, fallback } => {
+            let AbstractByte::Init(val, _) = bytes[offset.bytes()]
+                else { return None };
+            let next_discriminator = children.get(val).unwrap_or(fallback);
+            compute_discriminant::<M>(bytes, next_discriminator)
+        }
+    }
+}
+
 impl Type {
-    fn decode<M: Memory>(Type::Enum { .. }: Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
-        todo!()
+    fn decode<M: Memory>(Type::Enum { variants, discriminator, size, .. }: Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
+        if bytes.len() != size.bytes() { throw!(); }
+        let disc = compute_discriminant::<M>(bytes, discriminator)?;
+
+        // Decode into the variant.
+        // Because the variant is the same size as the enum we don't need to pass a subslice.
+        let Some(value) = variants[disc].ty.decode(bytes)
+            else { return None };
+
+        Some(Value::Variant { idx: disc, data: value })
     }
 
-    fn encode<M: Memory>(Type::Enum { .. }: Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
-        todo!()
+    fn encode<M: Memory>(Type::Enum { variants, .. }: Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
+        let Value::Variant { idx, data } = val else { panic!() };
+
+        // `idx` is guaranteed to be in bounds by the well-formed check in the type.
+        let Variant { ty: variant, tagger } = variants[idx];
+        let mut bytes = variant.encode(data.extract());
+
+        // Write tag into the bytes around the data.
+        // This is fine as we don't allow encoded data and the tag to overlap.
+        for (offset, value) in tagger.iter() {
+            assert_eq!(Some(AbstractByte::Uninit), bytes.get(offset.bytes()));
```

Comment:
This assertion is somewhat dubious. We should have the invariant that "if your code passes well-formedness then you will never get a panic", but here, this panic is not excluded by the WF-check.

So I think this should be just a comment.
```suggestion
            // If this enum type is "reasonable", we'll only overwrite `Uninit` here`.
            // However check_wf cannot ensure this, so we do not have an assertion.
```